rules/opentitan_test.bzl

```diff
@@ -187,6 +188,77 @@ def verilator_params(
     )
     return kwargs
 
+def qemu_params(
+        # Base Parameters
+        args = _BASE_PARAMS["args"],
+        data = _BASE_PARAMS["data"],
+        exit_success = _BASE_PARAMS["exit_success"],
+        exit_failure = _BASE_PARAMS["exit_failure"],
+        local = _BASE_PARAMS["local"],
+        otp = _BASE_PARAMS["otp"],
+        rom = _BASE_PARAMS["rom"],
+        tags = _BASE_PARAMS["tags"],
+        timeout = _BASE_PARAMS["timeout"],
+        test_runner = _BASE_PARAMS["test_runner"],
+        test_cmds = _BASE_PARAMS["test_cmds"] + [
+            "console",
+            "--exit-success={exit_success}",
+            "--exit-failure={exit_failure}",
+        ],
+        # QEMU-specific Parameters
+        aes_fast_mode = True,
+        # TODO add more QEMU OT parameters: https://github.com/lowRISC/qemu/blob/ot-earlgrey-8.0.2/docs/opentitan/earlgrey.md
+        # TODO handle SPI flash image
+        # None
+        **kwargs):
+    """A macro to create QEMU sim parameters for OpenTitan functional tests.
+
+    This macro emits a dictionary of parameters which are pasted into the
+    QEMU simulation specific test rule.
+
+    Parameters:
+        @param args: Extra arguments to pass to the test runner (`opentitantool`).
+        @param data: Data dependencies of the test.
+        @param local: Whether the test should be run locally without sandboxing.
+        @param otp: The OTP image to use.
+        @param rom: The ROM image to use.
+        @param tags: The test tags to apply to the test rule.
+        @param timeout: The timeout to apply to the test rule.
+        @param test_cmds: A list of required commands and args that make up the
+                          immutable portion of the harness invocation.
+        @param aes_fast_mode: Set to False to enable a more accurate but slower
```

Comment:
Yes I am aware of this. I put the FIXME because I am not sure how to handle the various options available. We could allow passing arbitrary qemu command-line options but this is quite ugly and fragile as well. I need to think about it more. It feels like some options like `icount` are important and special. Maybe the `-global` options could be passed as an array? e.g. `global_options = ["ot-aes.fast-mode", "false"]`?

Reply:
I'm afraid it may become more complex that the current somewhat "simple" examples (see `qemu-system-riscv32 -h` if you want to get really scared). I'm not sure whether `-global` is any special here.  Other options such as `-object`, `-device` etc. may follow the same path. I'm not sure if trying to control QEMU options from bazel can help - it could become a bottleneck. I believe you need to provide a way to forward any options to QEMU, as you would do for a compiler for example. This is why our scripts always use `--` as a std separator to forward options to QEMU that the script to not handle on its own. I'm not a Bazel user so I cannot really provide useful pieces of advice, though.

Reply:
For CI/testing, it is important that each test be able to specify whatever options are needed to run it. For tests that are run using `bazel run`, it is possible to give extra arguments that are forwarded to the runner (opentitantool in this case). So this following could be possible (and probably how it will be handled under the hood)
```bash
bazel run //sw/device/tests/crypto:aes_functest_sim_qemu --qemu-extra-opt="-global ot-aes.fast-mode=false"
```
However it obviously does not scale very well because if you further want to do this with `bazel test` it becomes:
```bash
bazel run //sw/device/tests/crypto:aes_functest_sim_qemu --test_arg=--qemu-extra-opt="-global ot-aes.fast-mode=false"
```
I think it's just more useable if the options are specified in the bazel. Based on what you are saying, it seems that:
- `icount` is special so should have its own entry
- machine properties cannot be trated as extra command line arguments because it has to be specified in one go I think (`-M ot-earlgrey,<insert machine props>`) so must have their own entry as well
- all other can be just extra arguments passed to QEMU?
So this would give:
```python
opentitan_functest(
   ...,
   qemu_param(
     icount = 6,
     machine_props = ["no_epmp_cfg=true"],
     extra_options = ["-global", "ot-aes.fast-mode=false"]
   )
```
However this should definitely be abstracted in some way: we don't want every test to mess with options like this. The way it is done currently is to define global variables:
```python
# somewhere in the bazel library:
QEMU_OT_AES_NO_FAST_MODE=["-global", "ot-aes.fast-mode=false"]

# when defining the test:
opentitan_functest(
   ...,
   qemu_param(
     icount = 6,
     machine_props = ["no_epmp_cfg=true"],
     extra_options = QEMU_OT_AES_NO_FAST_MODE,
   )
```
What do you think?

Reply:
It seems ok.